kvserver/rangefeed: add capacity to node level buffered sender

wenyihu6 opened the pull request:
Resolves: https://github.com/cockroachdb/cockroach/issues/152505
Release note: none

---
**kvserver/rangefeed: add capacity to node level buffered sender**

This patch adds capacity to node level buffered sender which will shut down all
registrations if the node level buffer had overflowed.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

---
**kvserver/rangefeed: rename overflow to overflowed**

This commit renames overflow to overflowed in BufferedSender for clarity.

---
**kvserver/rangefeed: dynamically adjust capacity per buffered sender**

Previously, buffered sender capacity was configured via a cluster setting. This
commit removes that setting and makes capacity scale dynamically with the number
of active registrations. The minimum (and default) is 4096 * 20, and capacity
never drops below that. Above the minimum, it grows as 4096 * <active
registrations>. Events already buffered will not dropped when capacity
decreases, and capacity is not adjusted during shutdown (0).

---
**rangefeed: move back to static (bug configurable) send queue limit**

The interactions here probably require more thought before doing
anything more complicated than a simple static limit.

Epic: none
Release note: None

---
**rangefeed: fix incorrect comment**

Epic: none
Release note: None

---
**rangefeed: deflake TestUnbufferedRegWithStreamManager**

Under stress, this test would fail because attempting to publish 20
events in a tight loop would overwhelm the processor's event buffer
(which has a capacity of 16).

Epic: none
Release note: None

wenyihu6 committed:
kvserver/rangefeed: add capacity to node level buffered sender

This patch adds capacity to node level buffered sender which will shut down all
registrations if the node level buffer had overflowed.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153740)
<!-- Reviewable:end -->

stevendanna reviewed (commented):
Thanks for picking this up!

I left some comments inline. I've gone back and forth on whether this should be a memory limit instead of a count limit. I think I like the simplicity of the count for now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-:-O_VzyDS6VgXUBZ1p1nX:b-ty9zwm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_Vz_Ph2lVEmv4BDwEO:-O_Vz_Pi-Najz9-pBU-G:b-or4pnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/replica_rangefeed.go#L102)):*
> ```Go
> 	// what should we set here a single buffered sender here for one Rangefeed
> 	// call
> 	2048,
> ```

If we go with a limit on the number of events, we probably want this to be some multiple of the old per-registration limit of 8192.  I could imagine something where we ratchet this up as new streams are added and then (carefully) ratchet it back down when they are removed.
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_W-PErC6VjviucYL3-:-O_W-PErC6VjviucYL30:b-zg2j9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/rangefeed/buffered_sender.go#L107)):*
> ```Go
> 		log.Dev.Error(context.Background(), "buffer capacity exceeded")
> 		return newRetryErrBufferCapacityExceeded()
> 	}
> ```

I wonder if we need something here to always allow error events?
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_W0Pyb35GL82voaauo:-O_W0Pyb35GL82voaaup:bysexta) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/rangefeed/buffered_sender.go#L161)):*
> ```Go
> 				}
> 				if overflowed && remains == int64(0) {
> 					return newRetryErrBufferCapacityExceeded()
> ```

I think we'll want a test or two to make sure this definitely makes it back to clients.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver/rangefeed: rename overflow to overflowed

This commit renames overflow to overflowed in BufferedSender for clarity.

wenyihu6 committed:
kvserver/rangefeed: dynamically adjust capacity per buffered sender

Previously, buffered sender capacity was configured via a cluster setting. This
commit removes that setting and makes capacity scale dynamically with the number
of active registrations. The minimum (and default) is 4096 * 20, and capacity
never drops below that. Above the minimum, it grows as 4096 * <active
registrations>. Events already buffered will not dropped when capacity
decreases, and capacity is not adjusted during shutdown (0).

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153740)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_W-PErC6VjviucYL3-:-O_mZ5xdB7HqOoAUhThd:b-iib3uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/rangefeed/buffered_sender.go#L107)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I wonder if we need something here to always allow error events?
</blockquote></details>

If we allow error events, they’ll just be buffered at the end of the queue. Once the queue is empty and the `overflowed` flag is set, the grpc stream will terminate https://github.com/cockroachdb/cockroach/blob/9f848afd2bacd3dc38ff1661e764ce29288a9c28/pkg/server/node.go#L2246-L2247. So clients may not receive explicit retry error events, but they will see the stream context cancelled - which will effectively be the same outcome as allowing error events. wdyt?
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_W0Pyb35GL82voaauo:-O_mZjTb20aXxJSIC7sF:b-8e16mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/rangefeed/buffered_sender.go#L161)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think we'll want a test or two to make sure this definitely makes it back to clients.
</blockquote></details>

I added some unit tests on the server side. `TestBufferedSenderOnStreamShutdown` tests that we do send something to the error channel https://github.com/cockroachdb/cockroach/blob/9f848afd2bacd3dc38ff1661e764ce29288a9c28/pkg/kv/kvserver/rangefeed/buffered_sender_test.go#L267 which should shut down grpc stream here https://github.com/cockroachdb/cockroach/blob/9f848afd2bacd3dc38ff1661e764ce29288a9c28/pkg/server/node.go#L2246-L2247. Lmk if you think we should add more to the client side.
___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_Vz_Ph2lVEmv4BDwEO:-O_m_FbV07V6RPHBuGQ0:b-g8raed) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/replica_rangefeed.go#L102)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

If we go with a limit on the number of events, we probably want this to be some multiple of the old per-registration limit of 8192.  I could imagine something where we ratchet this up as new streams are added and then (carefully) ratchet it back down when they are removed.
</blockquote></details>

Good idea, removed this as a cluster setting and dynamically adjust based on number of registrations. 


<!-- Sent from Reviewable.io -->

stevendanna committed:
rangefeed: move back to static (but configurable) send queue limit

The interactions here probably requires more thought before doing
anything more complicated than a simple static limit.

Epic: none
Release note: None

wenyihu6 reviewed (commented):
Your changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the question about whether 8 * DefaultRangefeedEventCap is enough - I can be convinced either way just wanted to double check. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-:-O_r1Tkq1u0U3xB6gyX5:b669ab7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Not a blocking comment btw - with branch cut coming up, feel free to merge as is. We can defer this discussion to later.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-:-O_t0rUj3PgiLA-_z_39:b-g6mj8l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
This new test we are adding here TestBufferedSenderOnStreamShutdown seems to be deadlocking under stress - https://teamcity.cockroachdb.com/repository/download/Cockroach_Ci_Tests_MaybeStress/20505700:id/bazel-testlogs/pkg/kv/kvserver/rangefeed/rangefeed_test/test.log. Perhaps it has something to do with using BlockSend https://github.com/cockroachdb/cockroach/blob/6dbe5e342c7866672ff5db56709f2d76c8f03a7c/pkg/kv/kvserver/rangefeed/buffered_sender_test.go#L254. 

```
goroutine 74 [semacquire, 18 minutes]:
sync.runtime_Semacquire(0x3d2ccf0?)
	GOROOT/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc000100708?)
	GOROOT/src/sync/waitgroup.go:118 +0x48
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*StreamManager).Stop(0xc00066ef50, {0x29cf868, 0x3ff8060})
	pkg/kv/kvserver/rangefeed/stream_manager.go:188 +0x5c
panic({0x21a1c80?, 0x3d2cba0?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x43e5a5?, {0x29cf868, 0x3ff8060})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x21a1c80?, 0x3d2cba0?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000932900, {0x29cf868, 0x3ff8060})
	pkg/util/stop/stopper.go:568 +0x165
panic({0x21a1c80?, 0x3d2cba0?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.TestBufferedS
```

stevendanna commented:
@wenyihu6 Thanks for the pointer. There were a few issues here, but the unblocking behaviour was definitely the place to start.

stevendanna commented:
If the tests pass I'll rebase and merge. We can debate and update the default in a follow up.  I think it is hard to set a principled default here.

stevendanna committed:
rangefeed: deflake TestUnbufferedRegWithStreamManager

Under stress, this test would fail because attempting to publish 20
events in a tight loop would overwhelm the processor's event buffer
(which has a capacity of 16).

Epic: none
Release note: None

stevendanna reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Let's keep an eye on this over the next few days, if it causes flakes, we can tune the limit or disable it by default.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-:-O_wbu7RBdnlq8hJi2H4:bp0qjwb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153740#-O_W-PErC6VjviucYL3-:-O_w_qVxBbpT3iqlxbvH:b-haeqry) ([raw file](https://github.com/cockroachdb/cockroach/blob/5536f0828f50bb21ec1577c77d388c4303d124a4/pkg/kv/kvserver/rangefeed/buffered_sender.go#L107)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

If we allow error events, they’ll just be buffered at the end of the queue. Once the queue is empty and the `overflowed` flag is set, the grpc stream will terminate https://github.com/cockroachdb/cockroach/blob/9f848afd2bacd3dc38ff1661e764ce29288a9c28/pkg/server/node.go#L2246-L2247. So clients may not receive explicit retry error events, but they will see the stream context cancelled - which will effectively be the same outcome as allowing error events. wdyt?
</blockquote></details>

Looking at the code, I think we have good coverage that sm.Error() should return the correct error and I believe that should end up being the one the client receives.  Let's merge this but I'll add a test that shows whatever the behaviour is in a follow up. 


<!-- Sent from Reviewable.io -->
